core/trie.go

```diff
@@ -0,0 +1,351 @@
+package core
+
+import (
+	"bytes"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/NethermindEth/juno/core/felt"
+	"github.com/bits-and-blooms/bitset"
+)
+
+type (
+	StoragePath  = bitset.BitSet
+	StorageValue = TrieNode
+)
+
+type (
+	TrieKey   = felt.Felt
+	TrieValue = felt.Felt
+)
+
+// Persistent storage backend for [Trie]
+type TrieStorage interface {
+	Put(key *StoragePath, value *StorageValue) error
+	Get(key *StoragePath) (*StorageValue, error)
+}
+
+// A [Trie] node
+type TrieNode struct {
+	value *TrieValue
+	left  *StoragePath
+	right *StoragePath
+}
+
+// Calculates Pedersen hash of a [TrieNode]
+func (n *TrieNode) Hash(specPath *StoragePath) *TrieValue {
+	if specPath.Len() == 0 {
+		return n.value
+	}
+
+	hash := new(TrieValue)
+	pathWords := specPath.Bytes()
+	if len(pathWords) > 4 {
+		panic("Path too long to fit in Felt")
+	}
+
+	var pathBytes [32]byte
+	for idx, word := range pathWords {
+		startBytes := 24 - (idx * 8)
+		binary.BigEndian.PutUint64(pathBytes[startBytes:startBytes+8], word)
+	}
+
+	pathFelt := felt.NewFelt(0)
+	(&pathFelt).SetBytes(pathBytes[:])
+
+	// https://docs.starknet.io/documentation/develop/State/starknet-state/
+	Pedersen(n.value, &pathFelt, hash)
+	pathFelt.SetUint64(uint64(specPath.Len()))
+	return hash.Add(hash, &pathFelt)
+}
+
+// Equality check of 2 [TrieNode] objects
+func (n *TrieNode) Equal(other *TrieNode) bool {
+	return n.value.Equal(other.value) && n.left.Equal(other.left) && n.right.Equal(n.right)
+}
+
+// Serializes a [TrieNode] in to a byte array
+func (n *TrieNode) MarshalBinary() ([]byte, error) {
+	var ret []byte
+	valueB := n.value.Bytes()
+	ret = append(ret, valueB[:]...)
+
+	if n.left != nil {
+		ret = append(ret, 'l')
+		leftB, err := n.left.MarshalBinary()
+		if err != nil {
+			return nil, err
+		}
+		ret = append(ret, leftB...)
+	}
+
+	if n.right != nil {
+		ret = append(ret, 'r')
+		rightB, err := n.right.MarshalBinary()
+		if err != nil {
+			return nil, err
+		}
+		ret = append(ret, rightB...)
+	}
+	return ret, nil
+}
+
+// Deserializes a [TrieNode] from a byte array
+func (n *TrieNode) UnmarshalBinary(data []byte) error {
+	if len(data) < felt.Bytes {
+		return errors.New("Malformed TrieNode bytedata")
+	}
+	n.value = new(TrieValue).SetBytes(data[:felt.Bytes])
+	data = data[felt.Bytes:]
+
+	stream := bytes.NewReader(data)
+	for stream.Len() > 0 {
+		head, err := stream.ReadByte()
+		if err != nil {
+			return err
+		}
+
+		var pathP **StoragePath
+		if head == 'l' {
+			pathP = &(n.left)
+		} else if head == 'r' {
+			pathP = &(n.right)
+		} else {
+			return errors.New("Malformed TrieNode bytedata")
+		}
+
+		*pathP = new(StoragePath)
+		_, err = (*pathP).ReadFrom(stream)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// State [Trie]
+type Trie struct {
+	root    *StoragePath
+	storage TrieStorage
+}
+
```

Comment:
I think we need a `NewTrie` function here so we can create Tries outside this package, since `storage` is private.
```suggestion

func NewTrie(storage TrieStorage) *Trie {
	return &Trie{
		storage: storage,
	}
}

```